File: 백준/Silver/11651. 좌표 정렬하기 2/좌표 정렬하기 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

// https://www.acmicpc.net/problem/11651
public class Main {
    private static int[][] sorted;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[][] arr = new int[N][2];
        sorted = new int[N][2];

        for (int i = 0; i < N; i++) {
            String[] inputs = br.readLine().split(" ");
            arr[i][0] = Integer.parseInt(inputs[0]);
            arr[i][1] = Integer.parseInt(inputs[1]);
        }

        mergeSort(arr, 0, N - 1);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < N; i++) {
            sb.append(arr[i][0])
                    .append(' ')
                    .append(arr[i][1])
                    .append('\n');
        }
        System.out.println(sb);
    }

    public static void mergeSort(int[][] arr, int left, int right) {
        if (left >= right) return;
        int mid = (left + right) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, right);
    }

    public static void merge(int[][] arr, int left, int right) {
        int i = left;
        int mid = (left + right) / 2;
        int j = mid + 1;
        int k = left;

        while (i <= mid && j <= right) {
            if (arr[i][1] < arr[j][1]) {
                sorted[k++] = arr[i++];
            } else if (arr[i][1] == arr[j][1]) {
                if (arr[i][0] <= arr[j][0]) {
                    sorted[k++] = arr[i++];
                } else {
                    sorted[k++] = arr[j++];
                }
            } else {
                sorted[k++] = arr[j++];
            }
        }

        for (; i <= mid; i++, k++) {
            sorted[k] = arr[i];
        }

        for (; j <= right; j++, k++) {
            sorted[k] = arr[j];
        }

        for (int l = left; l <= right; l++) {
            arr[l] = sorted[l];
        }
    }
}

File: 백준/Silver/11651. 좌표 정렬하기 2/좌표 정렬하기 2.cc
#include <stdio.h>
#include <stdlib.h>
int arr[2][100000];
int sorted[2][100000];

void merge(int** arr, int left, int mid, int right) {
	int i = left;
	int j = mid + 1;
	int k = left;

	while (i <= mid && j <= right) {
		if (arr[1][i] < arr[1][j]) {
			sorted[0][k] = arr[0][i];
			sorted[1][k++] = arr[1][i++];
		}
		else if (arr[1][i] == arr[1][j]) {
			if (arr[0][i] < arr[0][j]) {
				sorted[0][k] = arr[0][i];
				sorted[1][k++] = arr[1][i++];
			}
			else {
				sorted[0][k] = arr[0][j];
				sorted[1][k++] = arr[1][j++];
			}
		}
		else {
			sorted[0][k] = arr[0][j];
			sorted[1][k++] = arr[1][j++];
		}
	}
	for (; i <= mid; i++) {
		sorted[0][k] = arr[0][i];
		sorted[1][k++] = arr[1][i];
	}
	for (; j <= right; j++) {
		sorted[0][k] = arr[0][j];
		sorted[1][k++] = arr[1][j];
	}
	for (int l = left; l <= right; l++) {
		arr[0][l] = sorted[0][l];
		arr[1][l] = sorted[1][l];
	}
}
void merge_sort(int** arr, int left, int right) {
	int mid = (right + left) / 2;

	if (left < right) {
		merge_sort(arr, left, mid);
		merge_sort(arr,  mid + 1, right);
		merge(arr, left, mid, right);
	}
}
int main() {
	int N;

	scanf("%d", &N);
	int** arr = (int**)malloc(sizeof(int*) * 2);
	for (int i = 0; i < 2; i++) {
		arr[i] = (int*)malloc(sizeof(int) * N);
	}
	for (int i = 0; i < N; i++)
		scanf("%d %d", &arr[0][i], &arr[1][i]);


	merge_sort(arr, 0, N - 1);
	for (int i = 0; i < N; i++) {
		printf("%d %d\n", arr[0][i], arr[1][i]);
	}

	return 0;
}